### PR TITLE
prevent duplicate UUIDs from being created and allow users to update hostnames based on uuid

### PR DESCRIPTION
**Pull request by rebeccahhh**
##### SUMMARY

if the user provides a uuid and it exists, allow that to tie to the instance, which allows the user to update the instance based on the UUID (includeding updating the hostname) should they choose to do so.

##### ISSUE TYPE

 - Feature Pull Request
 - Bugfix Pull Request

##### COMPONENT NAME

 - API

##### AWX VERSION

```
awx: 19.3.0
```

##### ADDITIONAL INFORMATION

```
awx-manage provision_instance
```
##### TODO
---
- [ ] Modify the installer to make the `system_uuid` consistent (see discussion below)

### CODE REVIEW

File: awx/main/managers.py
Comment (AlanCoding) [inline awx/main/managers.py]: you should combine this with

https://github.com/ansible/awx/blob/6a17e5b65bf9421c2e650d23817c79ce81248c6f/awx/main/models/ha.py#L62

File: awx/main/managers.py
Comment (jbradberry) [inline awx/main/managers.py]: this should be written as `uuid is not None...`

File: awx/main/managers.py
Comment (rebeccahhh) [reply awx/main/managers.py]: you're right, good call.

File: awx/main/managers.py
Comment (rebeccahhh) [reply awx/main/managers.py]: I kept it this way to match the existing if statements below, but I'm not set on that.

**Review by sarabrajsingh — APPROVED** [review]

File: awx/main/managers.py
Comment (sarabrajsingh) [reply awx/main/managers.py]: i conditionally approved, pending these changes!

File: awx/main/managers.py
Comment (AlanCoding) [inline awx/main/managers.py:139]: Someone gives both hostname and uuid in the provision_instance command.

There is a hit based on hostname, but there is not a hit based on uuid. First, `instance` is defined to be a non-empty queryset. Then, it is redefined to be an empty queryset.

Later in the code, we find `if instance.exists()` is False and because of that, run `instance = self.create`. The Instance table is still, absolutely, unique on hostname, so it would give an IntegrityError if I read it right.

I could be wrong, but a unit test would be good. I don't know what your expectation is, but that has tie-ins with how the installer is written.

**Comment by AlanCoding** [comment]
@jladdjr has merged related installer changes and I'd like him to make sure this is compatible with what that does.

**Comment by tchellomello** [comment]
To provide context, please assume the following dataset
```
>>> Instance.objects.values('hostname', 'uuid')
<QuerySet [{'hostname': '192.168.111.108', 'uuid': '5ffbd38ae5734483882edbb68ee1a43a'},
 {'hostname': '192.168.111.115', 'uuid': '2b9b8b03-8de8-466a-8f9f-49b0283079f2'}, 
{'hostname': 'controller_p70', 'uuid': '241c61d9-14ca-45fb-be38-ca8d55894e08'},
 {'hostname': 'controller_t470n1', 'uuid': 'e98da768-6fb0-4195-b47d-c85aba7155df'},
 {'hostname': 'node_p70', 'uuid': 'e82814c6df6e4db6a93beb6c041f76a5'}]>
>>> Instance.objects.values('hostname', 'uuid').count()
5

```
Then testing:

```
[root@126-addr ~]# awx-manage  provision_instance --hostname='controller_p70' --node_type='control' --uuid='241c61d9-14ca-45fb-be38-ca8d55894e08' 
Instance already registered controller_p70
[root@126-addr ~]# awx-manage  provision_instance --hostname='controller_p70' --node_type='control' --uuid='241c61d9-14ca-45fb-be38-ca8d55NEW' 
Instance already registered controller_p70
[root@126-addr ~]# awx-manage  provision_instance --hostname='controller_p70_renamed' --node_type='control' --uuid='241c61d9-14ca-45fb-be38-ca8d55894e08' 
Successfully registered instance controller_p70_renamed
(changed: True)
[root@126-addr ~]# awx-manage  provision_instance --hostname='controller_p70_renamed' --node_type='control'
Instance already registered controller_p70_renamed
[root@126-addr ~]# awx-manage  provision_instance --hostname='controller_no_uuid' --node_type='control'
Successfully registered instance controller_no_uuid
(changed: True)
```

Which provides:
```
>>> Instance.objects.values('hostname', 'uuid').count()
6
>>> Instance.objects.values('hostname', 'uuid')
<QuerySet [{'hostname': '192.168.111.108', 'uuid': '5ffbd38ae5734483882edbb68ee1a43a'},
 {'hostname': '192.168.111.115', 'uuid': '2b9b8b03-8de8-466a-8f9f-49b0283079f2'}, 
{'hostname': 'controller_no_uuid', 'uuid': '00000000-0000-0000-0000-000000000000'}, 
{'hostname': 'controller_p70_renamed', 'uuid': '241c61d9-14ca-45fb-be38-ca8d55894e08'},
 {'hostname': 'controller_t470n1', 'uuid': 'e98da768-6fb0-4195-b47d-c85aba7155df'}, 
{'hostname': 'node_p70', 'uuid': 'e82814c6df6e4db6a93beb6c041f76a5'}]>
```

File: awx/main/managers.py
Comment (rebeccahhh) [reply awx/main/managers.py]: is that along the lines of what you were thinking @AlanCoding ?

File: awx/main/managers.py
Comment (rebeccahhh) [reply awx/main/managers.py]: updated, cheers y'all

File: awx/main/managers.py
Comment (jladdjr) [inline awx/main/managers.py]: It seems like the basic logic here is "default to looking up instance by uuid, if one is provided. otherwise, look up by hostname". Could make that more explicit with:

```
            # attempt to find instance by uuid
            if uuid is not None and uuid != UUID_DEFAULT and self.filter(uuid=uuid).exists():
                instance = self.filter(uuid=uuid)
            else:
                # otherwise, rely on hostname
                instance = self.filter(hostname=hostname)
```

Comment (jladdjr) [reply awx/main/managers.py]: `update the existing uuid` doesn't seem accurate, since we're not updating the uuid here.

File: awx/main/managers.py
Comment (jladdjr) [inline awx/main/managers.py:144]: Would add something like this to call out what's happening here:
`# if instance was retrieved by uuid and hostname has changed, update hostname`

Comment (jladdjr) [reply awx/main/managers.py:144]: Would add a `logger.warn` here so that the logs reflect the hostname change happened.

**Review by jladdjr — APPROVED** [review]
Makes sense to make the `provision_instance` code a bit more defensive, so that `uuid` wins over the `hostname` and we update the hostname if a new one is provided 👍

File: awx/main/managers.py
Comment (AlanCoding) [reply awx/main/managers.py]: what if there is a match based on both hostname & uuid.

In that case, you would prefer the uuid, but... you would change the hostname, and instances are unique on hostname. This would probably error.

This pattern (of creating fallback lookup fields in the midst of unique fields) and the errors that come with it are familiar to me from the inventory import `instance_id` field. If the precedence ever matters, then you're left with an IntegrityError.

File: awx/main/managers.py
Comment (jbradberry) [inline awx/main/managers.py]: I think I had started to say something about this previously but got distracted.  This should probably be written a bit more defensively, since it will still get here and will error if there are multiple `Instance` objects pulled in by the queryset.  The most obvious option would be to use `.first()` instead of `.get()`, but we could also do something like with the conflicting ip address detection towards the top of this method instead.

File: awx/main/managers.py
Comment (rebeccahhh) [reply awx/main/managers.py]: so currently this will change the hostname if it's different and both a UUID and hostname are provided, I can see why that would be problematic, the solution would be to alert the user and verify that they want to change the hostname, but I can also see that getting very annoying very quickly.

File: awx/main/managers.py
Comment (AlanCoding) [reply awx/main/managers.py]: oh yeah, looks great, sorry for the delayed reply.

File: awx/main/managers.py
Comment (AlanCoding) [inline awx/main/managers.py:134]: Before I approve, I want to echo a concern for which discussions happened in other channels.

My reading of the installer before this point was that they would generate a new uuid every run (different from real uuid if an existing host exists). I would like to be sure that doesn't happen for re-runs of the installer, otherwise I get nervous. What makes the _most_ sense to me is for the installer to read the `ha.py` file's SYSTEM_UUID value if it's an existing instance, and generate one only if it's a new instance.

At this moment in code, if `self.filter(hostname=hostname).exists()`, then it will result in an error. I'm fine with that because it should throw an error in that case.

**Review by AlanCoding — APPROVED** [review]

**Comment by AlanCoding** [comment]
My current understanding is that this is needed for a fix in conjunction with some other installer changes, and I'd like to see it move along.

File: awx/main/managers.py
Comment (tchellomello) [reply awx/main/managers.py:134]: @AlanCoding yes, that makes sense. 

As you mentioned, the installer does generate a new `system_uuid` on every run, but it will only be used if the `ha.py` does not exist, **however**, at the task https://github.com/ansible/automation-platform-collection/blob/devel/roles/receptor/tasks/main.yml#L156-L159, we consume the `system_uuid` which is where the problem could happen. 

We will modify the installer to set the `system_uuid` from the `ha.py` and when `ha.py` then generate it fresh.

Comment (AlanCoding) [reply awx/main/managers.py:134]: Yeah, if reading from `ha.py` isn't needed for the current fix, and is just a matter of code cleanup, then feel free to file and issue to rework it later and get the currently needed fixes in. I think we're fully on the same page now.

**Review by jladdjr — APPROVED** [review]

File: awx/main/managers.py
Comment (tchellomello) [reply awx/main/managers.py:134]: @rebeccahhh++  opened the PR https://github.com/ansible/automation-platform-collection/pull/179 which should take of this

File: awx/main/managers.py
Comment (rebeccahhh) [reply awx/main/managers.py:144]: added line 145 should address that @jladdjr :)

**Review by beeankha — APPROVED** [review]

**Pull request merged**
